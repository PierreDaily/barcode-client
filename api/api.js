import axios from "axios";
import { Tokens } from "./class";

const serverAddress =
  process.env["SERVER_ADDRESS"] === "development"
    ? "192.168.0.183"
    : "pierredaily.com";
const serverPort = process.env["NODE_ENV"] === "development" ? 3001 : 3000;
const sheme = process.env["NODE_ENV"] === "development" ? "http" : "https";
const serverURL = `${sheme}://${serverAddress}:${serverPort}`;
const auth = new Tokens();

const API = {
  _getNewAccessToken: async function(refreshToken) {
    try {
      return await API.post("/auth", {}, { refresh_token: refreshToken }).then(
        ({ data: { access_token } }) => access_token
      );
    } catch (err) {
      return null;
    }
  },

  _fetch: async (method, path, params, headers) => {
    const options = {
      headers: {
        Authorization: `Bearer ${auth.accessToken}`,
        "Content-Type": "application/json",
        ...headers
      },
      method,
      data: params
    };

    try {
      return await axios(`${serverURL}${path}`, options);
    } catch (err) {
      if (err && err.response && err.response.status === 401) {
        auth.accessToken = await API._getNewAccessToken(auth.refreshToken);

        if (auth.accessToken) {
          const updatedOptions = {
            ...options,
            headers: {
              ...options.headers,
              Authorization: `Bearer ${auth.accessToken}`
            }
          };
          return axios(`${serverURL}${path}`, updatedOptions);
        } else {
          throw new Error("can't get a new access token");
        }
      } else {
        throw err;
      }
    }
  },
  get: (path, headers) => API._fetch("GET", path, null, headers),
  post: (path, params, headers) => API._fetch("POST", path, params, headers),
  put: (path, params, headers) => API._fetch("PUT", path, params, headers)
};

export { API };
