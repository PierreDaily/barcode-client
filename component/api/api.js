import axios from "axios";

const serverAddress =
  process.env["NODE_ENV"] === "development" ? "192.168.0.183" : "5.135.186.3";
const serverPort = 3000;
const sheme = "http";
const serverURL = `${sheme}://${serverAddress}:${serverPort}`;

const API = {
  get: (path, headers) => {
    const options = {
      ...headers,
      "Content-Type": "application/json"
    };
    return axios.get(`${serverURL}${path}`, options);
  },

  post: (path, params, headers) => {
    const options = {
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        ...headers
      },
      method: "POST",
      data: params
    };

    return axios(`${serverURL}${path}`, options);
  },

  put: (path, params, headers) => {
    const options = {
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        ...headers
      },
      method: "PUT",
      data: params
    };

    return axios(`${serverURL}${path}`, options);
  }
};

export default API;
