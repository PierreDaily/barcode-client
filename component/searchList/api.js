import API from "../api/api";

const customAPI = {
  postBrand: name => {
    return API.post("/brand", { name }).then(({ data }) => data);
  },
  searchBrand: name =>
    name !== ""
      ? API.get(
          `/brand-search/${name}`,
          { name: name.toLowerCase() },
          { "Content-Type": "application/json" }
        ).then(({ data }) => data)
      : Promise.resolve([])
};

export default customAPI;
