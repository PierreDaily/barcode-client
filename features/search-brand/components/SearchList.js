import React, { useState } from "react";
import PropTypes from "prop-types";
import { FlatList, SafeAreaView, TouchableOpacity } from "react-native";
import { List, Searchbar } from "react-native-paper";
import styles from "../styles";
import API from "../api";
import logger from "../../../logger";
import getErrMsg from "../helpers";
import { FlatListItemSeparator } from "./FlatListItemSeparator";

const SearchList = ({ navigation, route }) => {
  const [brandName, setBrandName] = useState("");
  const [brandList, setBrandList] = useState([]);
  const navigateToItemDetails = item => {
    navigation.navigate("Item-details", {
      ...route.params,
      brandId: item.id,
      brandName: item.name
    });
  };

  let trimedBrandName = brandName.trim();

  return (
    <SafeAreaView style={styles.container}>
      <Searchbar
        placeholder="Search"
        onChangeText={value => {
          setBrandName(value);
          API.searchBrand(value.trim())
            .then(result => {
              if (result) {
                setBrandList(result);
              }
            })
            .catch(err => logger(err));
        }}
        value={brandName}
        testID="search-bar"
      />
      {trimedBrandName !== "" && (
        <TouchableOpacity
          onPress={() =>
            API.postBrand(trimedBrandName)
              .then(item => navigateToItemDetails(item))
              .catch(err => {
                logger(getErrMsg(err));
                alert(getErrMsg(err));
              })
          }
          testID="new-brand"
        >
          <List.Item title={`Add new brand : "${trimedBrandName}"`} />
          <FlatListItemSeparator />
        </TouchableOpacity>
      )}
      <FlatList
        data={brandList}
        ItemSeparatorComponent={FlatListItemSeparator}
        renderItem={({ item }) => (
          <TouchableOpacity onPress={() => navigateToItemDetails(item)}>
            <List.Item title={item.name} />
          </TouchableOpacity>
        )}
        keyExtractor={item => item.id.toString()}
      />
    </SafeAreaView>
  );
};

SearchList.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired
  }),
  route: PropTypes.shape({ params: PropTypes.object })
};

export { SearchList };
