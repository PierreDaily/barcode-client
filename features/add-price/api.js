import { API } from "../../api";

const customAPI = {
  postPrice: ({ branchId, discountIndex, itemId, totalPrice }) => {
    return API.post("/price", {
      branchId,
      discountIndex,
      itemId,
      totalPrice: totalPrice * 100
    }).then(({ data }) => data);
  }
};

export default customAPI;
