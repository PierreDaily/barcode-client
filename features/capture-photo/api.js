import { API } from "../../api";

const customAPI = {
  postItem: ({ barcode, barcodeType, brand, name, imgUri }) => {
    let data = new FormData();
    data.append("image", {
      uri: imgUri,
      name: "image.jpg",
      type: "image/jpeg"
    });
    data.append("barcode", barcode);
    data.append("barcode_type", barcodeType);
    data.append("name", name);
    data.append("brand", brand);

    return API.post(`/item`, data, {
      "content-type": `multipart/form-data`
    }).then(({ data }) => data);
  }
};

export default customAPI;
