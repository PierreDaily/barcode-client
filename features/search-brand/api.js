import { API } from "../../api";

const customAPI = {
  postBrand: name => {
    return API.post("/brand", { name }).then(({ data }) => data);
  },
  searchBrand: name =>
    name !== ""
      ? API.get(`/brand-search/${name}`).then(({ data }) => data)
      : Promise.resolve([])
};

export default customAPI;
