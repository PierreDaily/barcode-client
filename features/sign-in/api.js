import { API } from "../../api";

const customAPI = {
  authenticate: (email, password) =>
    API.post("/auth", {
      email,
      password
    }).then(val => val.data)
};

export { customAPI };
