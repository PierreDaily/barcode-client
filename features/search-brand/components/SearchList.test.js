import React from "react";
import SearchList from "./SearchList";

import {
  act,
  cleanup,
  render,
  fireEvent,
  waitForElement,
  flushMicrotasksQueue
} from "react-native-testing-library";
import renderer from "react-test-renderer";
import api from "../api";
jest.mock("./api");

let minProps;
const spy = jest.spyOn(api, "searchBrand").mockImplementation(() =>
  Promise.resolve([
    {
      id: 1,
      name: "nestle"
    },
    {
      id: 2,
      name: "coca cola"
    }
  ])
);

const spyPost = jest.spyOn(api, "postBrand").mockImplementation(() =>
  Promise.resolve({
    id: 3,
    name: "new brand"
  })
);

beforeEach(() => {
  minProps = {
    navigation: { navigate: jest.fn() },
    route: { params: { barcode: "123456789", type: "EAN" } }
  };
});

afterEach(() => {
  jest.clearAllMocks();
  cleanup();
});

test("renders correctly", async () => {
  const tree = renderer.create(<SearchList {...minProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("query the API when typing in the search bar", async () => {
  const query = "coca";
  const { getByTestId } = render(<SearchList {...minProps} />);
  const searchBar = getByTestId("search-bar");
  expect(searchBar).toBeTruthy();
  fireEvent(searchBar, "onChangeText", query);
  await act(async () => flushMicrotasksQueue());
  expect(spy).toHaveBeenCalledWith(query);
});

test("navigate to Item-details screen with correct props on click", async () => {
  const query = "zerty";
  const expectedRoute = "Item-details";
  const expectedValues = {
    barcode: "123456789",
    brandId: 1,
    brandName: "nestle",
    type: "EAN"
  };
  const { getByTestId, getByText } = render(<SearchList {...minProps} />);
  const searchBar = getByTestId("search-bar");
  expect(searchBar).toBeTruthy();
  fireEvent(searchBar, "onChangeText", query);
  expect(spy).toHaveBeenCalledWith(query);

  await act(async () => waitForElement(() => getByText("nestle")));
  expect(getByText("nestle")).toBeTruthy();

  fireEvent.press(getByText("nestle"));

  expect(minProps.navigation.navigate).toHaveBeenCalledWith(
    expectedRoute,
    expectedValues
  );
});

test("display ADD brand option when typing any brand name", async () => {
  const query = "abcd";
  const { getByTestId } = render(<SearchList {...minProps} />);
  const searchBar = getByTestId("search-bar");
  expect(searchBar).toBeTruthy();
  fireEvent(searchBar, "onChangeText", query);
  expect(spy).toHaveBeenCalledWith(query);
  const newBrand = getByTestId("new-brand");
  await act(async () => waitForElement(() => newBrand));
  expect(newBrand).toBeTruthy();
});

test("navigate to Item-details with the right paramater if click on new brand", async () => {
  const expectedRoute = "Item-details";
  const expectedValues = {
    barcode: "123456789",
    brandId: 3,
    brandName: "new brand",
    type: "EAN"
  };
  const query = "abcd";

  const { getByTestId } = render(<SearchList {...minProps} />);
  const searchBar = getByTestId("search-bar");
  expect(searchBar).toBeTruthy();
  fireEvent(searchBar, "onChangeText", query);
  const newBrand = getByTestId("new-brand");
  await act(async () => waitForElement(() => newBrand));
  fireEvent.press(newBrand);
  await act(async () => flushMicrotasksQueue());
  expect(spyPost).toHaveBeenCalled();
  expect(minProps.navigation.navigate).toHaveBeenCalledWith(
    expectedRoute,
    expectedValues
  );
});
