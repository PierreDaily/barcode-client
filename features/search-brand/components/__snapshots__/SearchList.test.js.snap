// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "padding": "4%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderRadius": 4,
        "elevation": 4,
        "flexDirection": "row",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 4,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={false}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      <View
        style={null}
      >
        <Text />
      </View>
    </View>
    <TextInput
      accessibilityRole="search"
      accessibilityTraits="search"
      allowFontScaling={true}
      autoFocus={true}
      keyboardAppearance="light"
      onChangeText={[Function]}
      placeholder="Search"
      placeholderTextColor="rgba(0, 0, 0, 0.54)"
      rejectResponderTermination={true}
      returnKeyType="search"
      selectionColor="#6200ee"
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "flex": 1,
            "fontSize": 18,
            "minWidth": 0,
            "paddingLeft": 8,
            "textAlign": "left",
          },
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          undefined,
        ]
      }
      testID="search-bar"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessibilityRole="button"
      accessibilityStates={
        Array [
          "disabled",
        ]
      }
      accessible={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 6,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            },
            Object {
              "opacity": 0.32,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        style={null}
      >
        <Text />
      </View>
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
